### ISSUE TITLE
Shadowing of types via Generic Parameter

### ISSUE THREAD

**Issue by Havvy** [Manishearth/rust-clippy#380]
```
struct Foo {}

struct<Foo> Bar { f: Foo }
```

Should warn that `Foo` in `Bar` shadows the `struct Foo` declared previously.

**Comment by samueltardieu** [comment]
In modern Rust, a test-case could be

```rust
struct Foo;
struct Bar<Foo> { f: Foo };
```

Also, with today's compiler structure, working with a HIR visitor should be enough to detect such shadowing.

@rustbot label +good-first-issue -E-hard -T-middle
@rustbot note test-case-example

**Comment by devodev** [comment]
I'd like to try and tackle this as my first contribution.

@rustbot claim
